lib/Sema/CSSimplify.cpp

```diff
@@ -2007,6 +2038,48 @@ ConstraintSystem::matchTypes(Type type1, Type type2, ConstraintKind kind,
         conversionsOrFixes.push_back(ConversionRestrictionKind::DeepEquality);
       break;
     }
+
+    // Opaque archetypes are globally bound, so we can match them for deep
+    // equality.
+    case TypeKind::OpaqueTypeArchetype: {
+      auto opaque1 = cast<OpaqueTypeArchetypeType>(desugar1);
+      auto opaque2 = cast<OpaqueTypeArchetypeType>(desugar2);
+      
+      assert(!type2->is<LValueType>() && "Unexpected lvalue type!");
+      if (!type1->is<LValueType>()
+          && opaque1->getOpaqueDecl() == opaque2->getOpaqueDecl()) {
+        conversionsOrFixes.push_back(ConversionRestrictionKind::DeepEquality);
+      }
+      break;
+    }
+    
+    // Same for nested archetypes rooted in opaque types.
+    case TypeKind::NestedArchetype: {
+      auto nested1 = cast<NestedArchetypeType>(desugar1);
```

Comment:
Do you actually check anywhere that the two nested types are the same? It looks like you're only comparing the root types, but I might be wrong

Reply:
I check that the opaque decls match and that `interfaceTy1 == interfaceTy2` right below this.